Generic Mac build script for branded version of Mac manager/screen saver

armstrdj opened the pull request:
@CharlieFenton here is the current script. I still need to look at your suggestion for PostInstall.cpp to fix finding the branding file in the installer. Also I need to add support for setting up an account manager in the script. We have not finalized all the images/content for the WCG branded installer/manager so I have not checked in all of those images yet. 

Also I am still looking into some issues that I think are unrelated to the installer but may be issues in general with the manager or installer code.
1. The snooze/disconnect image is not showing up over the WCG icon in the taskbar.
2. The logo in the upper right notification bar uses the appropriate WCG logo defined in the skin. However if I switch the skin to default it properly switches the logo to boinc but does not switch the logo back when switching back to WCG skin.
3. The BOINC Installer icon displays in the screen saver popup during install instead of the bundle branded installer icon.

CharlieFenton commented:
@armstrdj I just finished the PR I have been working on. It includes a change to properly get the branding from the resources in the PostInstall.app, so be sure to put a copy there as well as in the BOINC Data directory.

I will review the contents of this PR and get back to you with my suggestions in the next day or two. Do I have write access to this branch, or will I need to email you any changes? 

I may already see one problem: you have not specified the wcg_play.png, wcg_pause.png or wcg_stop.png files in the skins.xml file. Please see [this description of the skins file](http://boinc.berkeley.edu/wiki/Creating_a_skin_for_the_BOINC_Manager).

armstrdj commented:
@CharlieFenton I fixed the branding file to be copied to the PostIntall.app and added the copying of the new installer icon to replace the icon in PostInstall.app which fixes my third issue above. For the skin I have several changes to the xml and will commit those soon. Adding the pause, play, and stop icons does fix those showing up in the upper notification bar but I think it is a different mechanism which is used for the taskbar icon.

I have the box checked to allow edits form the maintainers which I think should give you write access. If I am wrong and there is something else I need to do to give you write access let me know.

CharlieFenton commented:
> I fixed the branding file to be copied to the PostIntall.app

I merged master into this branch (so the files in this branch are up to date) because you will need my recent update to the PostInstall.cpp file. That update allows the PostInstall.app to find the branding file your script now puts there.

CharlieFenton commented:
@armstrdjI want to clear up some potential confusion about nomenclature. On the Mac, the row of icons (usually) at the bottom of the screen are called the _Dock_. The row of menus at the top of the screen is called the _menubar_. BOINC's 16 X 16 _play_, _pause/snooze_ and _disconnected_ icons in the right hand portion of the Mac menubar containing small icons with pull-down menus are called the _status bar_ in the Mac world. But since they are the Mac equivalent of the taskbar icon on Windows, we in the world of BOINC generally refer to those as BOINC's taskbar icons, even on the Mac.

So I think you are saying that you have now fixed the issue with the 16 X 16 icons in the Mac status bar but you are not yet seeing the pause/snooze or disconnected indicators in the Dock. I'll look into that, but I find it surprising because of the way I implemented those for the Dock. The Dock always shows the application icon (the same icon from the *.icns file that is shown in the Finder for the application.) The BOINC code superimposes the pause/snooze or disconnected symbol on top of that icon as a badge (in Mac terminology) when appropriate, so nothing special should be required for that to work with a branded icon.

Once you have committed you skin file changes, I'll try building the WCG branded installer so I can look into this and any other issues you find.

CharlieFenton commented:
@armstrdj I suggest that instead of requiring both a *.environment file and a separate *.branding file for each brand you just add a line to the environment file like this:
`BRANDING_INFO="BrandId=4"`
Then in the release_brand.sh script, replace these lines:
```
cp -fpRL ./mac_installer/${BRANDINGFILE} "../BOINC_Installer/Pkg_Root/Applications/${MANAGERAPPNAME}.app/Contents/Resources/Branding"
cp -fpRL ./mac_installer/${BRANDINGFILE} ../BOINC_Installer/Pkg_Root/Library/Application\ Support/BOINC\ Data/Branding

``` 
with these:
```
echo ${BRANDING_INFO} > "../BOINC_Installer/Pkg_Root/Applications/${MANAGERAPPNAME}.app/Contents/Resources/Branding"
echo ${BRANDING_INFO} > ../BOINC_Installer/Pkg_Root/Library/Application\ Support/BOINC\ Data/Branding

```
and replace this line:
`cp -fpRL ./mac_installer/${BRANDINGFILE} "../BOINC_Installer/New_Release_$1_$2_$3/${SHORTBRANDNAME}_$1.$2.$3_macOSX_$arch/${INSTALLERAPPNAME}.app/Contents/Resources/PostInstall.app/Contents/Resources/Branding"`
with this:
`echo ${BRANDING_INFO} > "../BOINC_Installer/New_Release_$1_$2_$3/${SHORTBRANDNAME}_$1.$2.$3_macOSX_$arch/${INSTALLERAPPNAME}.app/Contents/Resources/PostInstall.app/Contents/Resources/Branding"`

If at some point in the future we decide to put more extensive information in the Branding files, we can then decide whether to again pass a prebuilt Branding file to the script or just add more lines to the environment file.

CharlieFenton commented:
@armstrdj I would request that the script not put temporary files in the local repository. The release_boinc.sh script, upon which your script is based, creates temporary directories in the BOINC_Installer directory, which is the destination for the script's products but is not part of the local repository. I think it would be better practice to do the same with other temporary files.

I suggest changing this:
```
cp -fp mac_installer/complist.plist mac_installer/complist.plist.${SHORTBRANDNAME}
sed -i "" s/BOINCManager.app/"${MANAGERAPPNAME}.app"/g mac_installer/complist.plist.${SHORTBRANDNAME}
sed -i "" s/BOINCSaver.saver/"${SSAVERAPPNAME}.saver"/g mac_installer/complist.plist.${SHORTBRANDNAME}
cp -fp mac_installer/complist.plist.${SHORTBRANDNAME} ../BOINC_Installer/Installer\ templates/complist.plist

``` 
to this:
```
cp -fp mac_installer/complist.plist ../BOINC_Installer/Installer\ templates/complist.plist
sed -i "" s/BOINCManager.app/"${MANAGERAPPNAME}.app"/g ../BOINC_Installer/Installer\ templates/complist.plist
sed -i "" s/BOINCSaver.saver/"${SSAVERAPPNAME}.saver"/g ../BOINC_Installer/Installer\ templates/complist.plist

```
```
and similarly changing this:
cp -fp mac_installer/myDistribution mac_installer/myDistribution.${SHORTBRANDNAME}
sed -i "" s/BOINCManager.app/"${MANAGERAPPNAME}.app"/g mac_installer/myDistribution.${SHORTBRANDNAME}
sed -i "" s/BOINCSaver.saver/"${SSAVERAPPNAME}.saver"/g mac_installer/myDistribution.${SHORTBRANDNAME}
sed -i "" s/"BOINC Manager"/"${MANAGERAPPNAME}"/g mac_installer/myDistribution.${SHORTBRANDNAME}
cp -fp mac_installer/myDistribution.${SHORTBRANDNAME} ../BOINC_Installer/Installer\ templates/myDistribution

```
to this:
```
cp -fp mac_installer/myDistribution ../BOINC_Installer/Installer\ templates/myDistribution
sed -i "" s/BOINCManager.app/"${MANAGERAPPNAME}.app"/g ../BOINC_Installer/Installer\ templates/myDistribution
sed -i "" s/BOINCSaver.saver/"${SSAVERAPPNAME}.saver"/g ../BOINC_Installer/Installer\ templates/myDistribution
sed -i "" s/"BOINC Manager"/"${MANAGERAPPNAME}"/g ../BOINC_Installer/Installer\ templates/myDistribution

```

armstrdj commented:
@CharlieFenton I have made the changes you suggested and checked in some changes to the skin. I don't have the final products yet for the screen saver preview or the readme file so to build in the current state you will have to change the environment variables to:
SSAVERPREVIEW="boinc.jpg"
SSAVERTHUMBNAIL="thumbnail"
READMEFILE="ReadMe.rtf"

A couple of questions about the skin:
1. Off the top of your head do you know what the tag <is_branded> does? If not I can go look through the code.
2. There is some code in DlgAbout.cpp which uses the logo defined in the skin for the about window but it has been disabled. Is there and alternate method to use the branded logo on the about window?
#if 0
    m_AboutBOINCLogoCtrl->SetBitmap(wxBitmap(*(pSkinAdvanced->GetApplicationLogo())));

    m_AboutBOINCSloganCtrl->SetLabel(wxEmptyString);

    m_AboutBOINCURLCtrl->SetLabel(
        pSkinAdvanced->GetOrganizationWebsite().c_str()
    );
#endif

For the badge on the taskbar issue I do sometimes see the badge superimposed on the icon when it is suspended but most of the time it is not displayed. I am not sure the cause of when it appears and when it doesn't but if I figure anything out I will let you know.

CharlieFenton commented:
> Off the top of your head do you know what the tag <is_branded> does? 

The skin manager function _IsBranded()_ returns true if the tag indicates true. From looking at the code, it appears to be used by the Attach Wizard's _AccountInfoPage.cpp_ and _CompletionPage.cpp_ to set the branded name in the text of those wizard pages when attaching to branded account managers (GridRepublic, etc.) 

It looks to me that it would not be applicable for WCG, but it might be useful in case we find other dialogs which need branding. However, the way the code is now written, that might cause the _Attach to Account Manager_ wizard to incorrectly tell people they are attaching to WCG, so perhaps we should instead allow for 2 different tags, such as <is_branded> and <is__account_manager>, adjusting the wizard code accordingly. If the account managers want to build new branded installers, they will need to update their skins anyway; I think the most recent branded Mac installers were version 7.0.77 built in December 2013 though I'm not sure.

> There is some code in DlgAbout.cpp which uses the logo defined in the skin for the about window but it has been disabled. Is there and alternate method to use the branded logo on the about window?

GIT Blame shows the `#if 0` was added by @romw on 4/16/13, with the comment:
> MGR: Cleanup up the about dialog a little bit.

I suggest enabling it and see if that does what you want. 

> For the wizard section there is a title tag for the attach wizard which displays as the window title. For the account manager wizard there does not seem to be a title tag it looks like it just uses the same title as the attach wizard. Is there a way to have the two wizards have different window titles?

See my response about the <is_branded> tag earlier in this comment. You can probably use similar logic to accomplish what you want, testing `if (IS_ACCOUNTMANAGERWIZARD())`.

CharlieFenton commented:
> so perhaps we should instead allow for 2 different tags, such as <is_branded> and <is__account_manager>, adjusting the wizard code accordingly.

On second thought, rather than <is__account_manager>, I suggest <is__branded_account_manager> to avoid confusion with `if (IS_ACCOUNTMANAGERWIZARD())`.

CharlieFenton commented:
@armstrdj You may want to check whether the standard BOINC Copyright file (in the root directory of the repository) needs any adjustment for the WCG branded builds.
`cp -fp COPYRIGHT ../BOINC_Installer/New_Release_${SHORTBRANDNAME}_$1_$2_$3/${SHORTBRANDNAME}_$1.$2.$3_macOSX_$arch/extras/COPYRIGHT.txt`

CharlieFenton commented:
@armstrdj I also made the following change:
**lib/procinfo_mac.cpp**:
Recognize "World Community Grid.app" as part of BOINC when calculating BOINC and non-BOINC system resource usage

armstrdj commented:
@CharlieFenton Thanks for making these changes I will test today. For the icons we are developing new images to be used but they are not available yet so they will be replaced in the future.

CharlieFenton commented:
@armstrdj There is a problem using the same string for the _application_name_ and _application_short_name_ in your skin:
```
      <application_name>World Community Grid</application_name>
      <application_short_name>World Community Grid</application_short_name>
```
BOINC's _Help_ menu uses the short name for the first menu item (for general information about the software), and the full name for the second (for specific information on the user interface of the Manager.) If the short and full names are the same, these two menu items will have identical labels for a release (not pre-release) build. In general, I think it is important to distinguish between the general name of the software (BOINC, GridRepublic, etc.) and the GUI Manager application (BOINC Manager, GridRepublic Desktop.)

I assume you have dealt with this issue on the WCG branded Manager for MS Windows, but I have never used that so I don't know how you have handled it there.

armstrdj commented:
@CharlieFenton For the previous Windows versions we have used "World Community Grid - BOINC Manager" as the application name but are looking at changing it to just World Community Grid. How do you feel about changing the application name help to just be "Advanced view help" on the av and "Simple view help" on the sg? 

There are dialogs which use both the app name and short name together but I think in all cases the short name is referring to the client and is followed by client. If we were to keep the same string for both I will need to verify that for all cases.

CharlieFenton commented:
>  How do you feel about changing the application name help to just be "Advanced view help" on the av and "Simple view help" on the sg?

I would strongly discourage that. At least on the Mac, the standard for _Help_ menu items seems to be _{application name} Help_, though this seems less consistent on MS Windows. And I always think very carefully before changing any localized strings (which these are) because it can take quite some time for the volunteer translators to update all the languages. I think this may have changed, but I believe it was be true that even a single untranslated string invalidated the entire language translation. In any case, it is probably too late in the release process to change this in time for the 7.12 series releases. 

I agree that _World Community Grid - BOINC Manager_ is a bit long and clumsy. But we have always had the convention that the short and full application names are different and have somewhat different implications. and I feel that changing that convention makes this project a bigger issue which requires discussion among the BOINC Community about the possibility to make changes in the layout and structure of the skin.xml file.

> I think in all cases the short name is referring to the client and is followed by client.

Sorry, that is not so. I looked, and it is sometimes used for the client and sometimes used for the overall software (i.e., the combination of the client and Manager, and perhaps also the attached projects' applications.) And in cases where that is true, it is somewhat incorrect because the client's name  is never changed for branding; it is **always** named _boinc_.

**Examples of where it is used for the overall software or for the Manager:** 
In _AdvancedFrame.cpp_, it is used for menu items or their descriptions: `Exit %s` (when _is_boinc_started_by_manager_ is _true_), `%s &help`, `Show information about %s`, `%s &web site`, `See more information about %s on the web`, `Check for new %s version` and `%s has successfully added %s` (when syncing to an account manager).
In _sg_BOINCSimpleFrame.cpp_ it is used for menu items or their descriptions: `%s &help`, `Show information about %s`, `%s &web site`, `Check for new %s version`, `Get help with %s`
In _BOINCGUIApp.cpp_: `You currently are not authorized to manage %s.\n\nTo run %s as this user, please:\n- reinstall %s answering \"Yes\" ...`, `%s ownership or permissions are not set properly; please reinstall %s.\n(Error code %d`
In _BOINCTaskBar.cpp_: `Open %s Web...` (when using an account manager)

**Examples of where it is used for the client:**
In _BOINCBaseFrame.cpp_: `%s is not able to connect to a %s client.\nWould you like to try to connect again?`, `%s is not able to start a %s client.\nPlease launch the Control Panel->Administative Tools->Services applet and start the BOINC service."`, `%s is not able to start a %s client.\nPlease run the client and try again.`,` %s is not currently connected to a %s client.\nPlease use the 'Advanced\\Select Computer...' menu option to connect up to a %s client. ...`
_In BOINCClientManager.cpp_: `The %s client has exited unexpectedly 3 times within the last %d minutes.\nWould you like to restart it again?`

Our general policy has been to mention the concept of a client vs Manager as little as possible. So the following technically refer to the client, but are intentionally written to refer to the overall software:
In _BOINCDialupManager.cpp_: `%s needs to connect to the Internet.\nMay it do so now?`, `%s is connecting to the Internet.`, `%s has successfully connected to the Internet.`, `%s failed to connect to the Internet.`, `%s has detected it is now connected to the Internet.\nUpdating all projects and retrying all transfers.`, `%s has successfully disconnected from the Internet.`
in _DialogAdvPreferences.cpp_:`%s - Computing preferences` (window title), `If you run several projects, %s may switch between them this often.`, `Example: %s should transfer at most 2000 MB of data every 30 days.`, `%s will use the most restrictive of these settings:` and six more.
In _sg_DialogPreferencesBase.cpp_: `Limit the total amount of disk space used by %s.`, `Get help with %s`, `%s - Computing Preferences`(window title)
In _DlgDiagnosticLogFlags_: `%s Diagnostic Log Flags`(window title)
In _DlgExclusiveApps.cpp_: `%s - Exclusive Applications`(window title)
In _DlgExitMessage.cpp_: `This will shut down %s and its tasks until either the\n%s or the %s screen saver is run again. ...`
In _DlgHiddenColumns_: `%s Column Selection`(window title), `Select which columns %s should show.`
In _NoticeListCtrl.cpp_: `Notice from %s` (if notice is from the client)

CharlieFenton commented:
@armstrdj 
> Is there a way to have the two wizards have different window titles?

Since this was still not working, I debugged it. Apparently at one point the code to use the value of <attach_to_account_manager><title> was lost. I suspect this happened when Rom did a major overhaul of the Attach Wizard code. I have now added code to support this, and have committed it to this branch.

CharlieFenton commented:
Referring to translatable strings I wrote:
> In any case, it is probably too late in the release process to change this in time for the 7.12 series releases.

@armstrdj: If you want any chance of being able to release a branded build of version 7.12.x, we need to get this branch merged very soon, since it includes the changes to _clientgui/BOINCTaskBar.cpp_ and _clientgui/DlgAbout.cpp_ required for branding, and the changes to _SkinManager.cpp_, _SkinManager.h_ and _WizardAttach.cpp_, which enable use of the <attach_to_account_manager><title> tag in the skin.

I have also (separately) developed code which refactors the Mac branding code logic so that the strings associated with the various brands are all defined in only one source file. But I want to add that under a separate PR, which I can't create until this PR has been merged.

Your remaining changes to the WCG skin can be done under a subsequent PR, so you can ask this PR to be merged once you are happy with the way the script and new branding code is working.

armstrdj commented:
@CharlieFenton I am happy with the changes thus far. I don’t have the updated images yet but agree that it makes sense to merge this now to get all of the code changes in. I have not tested your latest change for the wizard title and won’t have connectivity again until late Friday to test. If you have tested it I am good with merging now.

CharlieFenton commented:
@davidpanderson @RichardHaselgrove or any other commiter: Could one of you please merge this PR into master. I will leave it to @armstrdj to say whether he hopes to create a WCG branded release of 7.12; if so, then this mere would also need to be ported to the 7.12 branch. Thank you.

RichardHaselgrove commented:
Hi @CharlieFenton  - I've been looking at this and mulling it over. On the one hand, I have a merge button, the automatic checks have passed, and I have the capacity to release the stalled development process.

But on the other hand, I know I'd be flying blind, I have no capacity to test the work so far, and I'd be taking it entirely on trust that your changes do what they say and don't introduce any collateral damage.

The whole process review we went through last year was designed to eliminate that 'taking on trust' of even the most experienced developer: I'm still waiting to see how my WSL fix #2581 is going to be handled, given that the main commit was branched into 7.12 but the fix wasn't. As a new and still inexperienced committer, I don't think I don't think I should be the one to relax policy: since the work here primarily affects World Community Grid, perhaps @TheAspens would have better knowledge of the context and could better assess it, if @davidpanderson doesn't feel able to do it?

CharlieFenton commented:
@RichardHaselgrove and all: Just to clarify, although this was developed by @armstrdj and me for the near-term use by World Community Grid, it is designed to be useful for any branded builds of BOINC, such as GridRepublic and CharityEngine.

JuhaSointusalo commented on mac_installer/release_brand.sh:
```diff
@@ -0,0 +1,400 @@
+    # XCode 3.x and 4.x use different paths for their build products.
+    # Our scripts in XCode's script build phase write those paths to 
+    # files to help this release script find the build products.
+    if [ "$4" = "-dev" ]; then
```
This doesn't work right. `$4` is the branding file.

JuhaSointusalo commented on mac_installer/release_brand.sh:
```diff
@@ -0,0 +1,400 @@
+    arch="universal"
+
+    # XCode 3.x and 4.x use different paths for their build products.
+    if [ "$4" = "-dev" ]; then
```
This doesn't work right. `$4` is the branding file.

JuhaSointusalo commented on mac_installer/release_brand.sh:
```diff
@@ -0,0 +1,400 @@
+if [ $# -lt 4 ]; then
+echo "Usage:"
+echo "   cd [path]/boinc"
+echo "   source [path_to_this_script] major_version minor_version revision_number brand_description_file"
```
This script changes enough environment variables that I _very strongly_ suggest not using `source`. If there is any concern about the script file not having execute permission set you can run the script with `bash script.sh`. Note that `source` is mentioned in the comment above too. Also note that the example line below doesn't use `source` due to lack of space after `.`.

JuhaSointusalo commented on mac_installer/release_brand.sh:
```diff
@@ -0,0 +1,400 @@
+#!/bin/csh
```
The script doesn't look like `csh` to me, though I can be wrong about that.

JuhaSointusalo commented on mac_installer/release_brand.sh:
```diff
@@ -0,0 +1,400 @@
+
+# This file is part of BOINC.
+# http://boinc.berkeley.edu
+# Copyright (C) 2008 University of California
```
Copyright year.

CharlieFenton commented:
@JuhaSointusalo Good catches, thank you. I'll fix these soon, when I have a chance. But I'm not sure I agree with you about not using `source`. I'll think abut that.

JuhaSointusalo commented:
> But I'm not sure I agree with you about not using `source`.

Well, I wouldn't use `source`. By design it modifies current shell, environment and shell options. That could lead to surprises later on. I'm not aware of any gotchas with `bash script.sh`. But at the end of the day, it's just a suggestion. If you want to use `source` then that's what you use.

armstrdj commented:
" I will leave it to @armstrdj to say whether he hopes to create a WCG branded release of 7.12" @CharlieFenton if we can get these changes in 7.12 that would be good.

JuhaSointusalo commented:
I was testing this (as much as I can test it on Windows) and noticed that About dialog says

> BOINC is distributed under the GNU Lesser General Public License v3.0.

even with WCG skin in use. Looking at source that text is translated but not branded. What should be done with it?
- nothing
- change now
- change later

armstrdj commented:
@CharlieFenton for the issue above in the About dialog box I am fine with looking at this in a later change if you are as well.

CharlieFenton commented:
> Looking at source that text is translated but not branded. What should be done with it?

I committed a fix. Thank you for the catch. I'm surprised you were able to test this under MS Windows.

JuhaSointusalo commented:
> I'm surprised you were able to test this under MS Windows.

Well, only the parts I could test, about dialog and attach wizard. For the rest, they looked ok and I'll just take your word that they work. @RichardHaselgrove had some concern about not being able to test this but until we get more Mac developers the best we can do is carefully read through any Mac changes.

JuhaSointusalo commented:
About `source`. What I had in mind is something like this.

a.sh:
```
#!/bin/sh
qqq=111
this --foo $qqq
```

b.sh:
```
#!/bin/sh
if [ "$1" = "x"]; then
    qqq=222
fi
that --bar $qqq
```

Now what `source ./b.sh` does depends on whether you had sourced `a.sh` before. It's a bug in `b.sh` all right but debugging it can be fun if you are like me and only close terminal windows when OS update requires reboot, with shell environment gathering who knows how much gunk.
